src/components/Form/types.ts

```diff
@@ -0,0 +1,77 @@
+import type {FocusEvent, Key, MutableRefObject, ReactNode, Ref} from 'react';
+import type {GestureResponderEvent, NativeSyntheticEvent, StyleProp, TextInputFocusEventData, ViewStyle} from 'react-native';
+import type {OnyxFormKey, OnyxValues} from '@src/ONYXKEYS';
+import type Form from '@src/types/onyx/Form';
+import type {BaseForm, FormValueType} from '@src/types/onyx/Form';
+
+type ValueTypeKey = 'string' | 'boolean' | 'date';
+
+type MeasureLayoutOnSuccessCallback = (left: number, top: number, width: number, height: number) => void;
+
+type BaseInputProps = {
+    shouldSetTouchedOnBlurOnly?: boolean;
+    onValueChange?: (value: unknown, key: string) => void;
+    onTouched?: (event: GestureResponderEvent) => void;
+    valueType?: ValueTypeKey;
+    value?: FormValueType;
+    defaultValue?: FormValueType;
+    onBlur?: (event: FocusEvent | NativeSyntheticEvent<TextInputFocusEventData>) => void;
+    onPressOut?: (event: GestureResponderEvent) => void;
+    onPress?: (event: GestureResponderEvent) => void;
+    shouldSaveDraft?: boolean;
+    shouldUseDefaultValue?: boolean;
+    key?: Key | null | undefined;
+    ref?: Ref<BaseInputProps>;
+    isFocused?: boolean;
+    measureLayout?: (ref: unknown, callback: MeasureLayoutOnSuccessCallback) => void;
+    focus?: () => void;
+};
+
+type InputWrapperProps<TInput, TInputProps extends BaseInputProps> = TInputProps & {
+    InputComponent: TInput;
+    inputID: string;
+};
```

Comment:
We should rethink those two types `BaseInputProps` and `InputWrapperProps`. I suggested in my first review to do something like this:

```ts
// src/components/Form/types.ts
type ValidInput = ElementType;

type InputProps<TInput extends ValidInput> = ComponentProps<TInput> & {
    shouldSetTouchedOnBlurOnly?: boolean;
    onValueChange?: (value: unknown, key: string) => void;
    onTouched?: (event: unknown) => void;
    valueType?: ValueType;
    onBlur?: (event: FocusEvent | Parameters<NonNullable<ComponentProps<TInput>['onBlur']>>[0]) => void;
};
```

This would allows us to use `InputWrapper` in any component and also provide type safety and intelisense when using the props of that component. With the current way I'm currently able to put any props I want in the `InputWrapper` without raising error:

```ts
            <InputWrapper
                InputComponent={TextInput}
                inputID="addressLine2"
                label="asass"
                aria-label="assaas"
                role={CONST.ACCESSIBILITY_ROLE.TEXT}
                defaultValue="ssd"
                maxLength={CONST.FORM_CHARACTER_LIMIT}
                spellCheck={false}
                asasasas={false} // wrong prop but TS is not complaining
                shouldSaveDraft={false}
            />
```

Also, we lost intelisense to suggest props from `TextInput`, e.g. `spellCheck`:

<img width="640" alt="Screenshot 2024-01-18 at 13 00 36" src="https://github.com/Expensify/App/assets/20051562/663b5c96-4525-4c48-86d4-2945d4f815db">

Reply:
We tried ElementType approach, but we failed 😅